File: ios/DesMoinesInsider/Views/Discover/SwipeCardStack.swift
import SwiftUI

/// Card-stack swipe deck. Renders the top 3 cards from `items` and exposes
/// callbacks for each commit direction. The parent ViewModel is the source
/// of truth for which item is on top — this view only owns the transient
/// drag offset for the in-flight gesture.
///
/// `containerWidth` and `containerHeight` must be supplied by the parent via
/// a `GeometryReader` measured *after* any padding is applied. Passing the
/// already-constrained size avoids the race where a background GeometryReader
/// could read the full unpadded screen width before padding resolves, which
/// caused cards to render wider than the viewport.
struct SwipeCardStack: View {
    let items: [SwipeItem]
    /// Available width of the deck area, measured after padding by the parent.
    let containerWidth: CGFloat
    /// Available height of the deck area, measured after padding by the parent.
    let containerHeight: CGFloat
    var onLike: (SwipeItem) -> Void
    var onSkip: (SwipeItem) -> Void
    var onBoost: (SwipeItem) -> Void
    var onTap: (SwipeItem) -> Void
    /// Set by the parent's action-bar buttons to drive the same animated fly-off
    /// as a gesture swipe (IOS-AUDIT-UX-018). Reset to nil once consumed.
    @Binding var command: Command?

    /// A button-driven swipe, mapped to the matching commit direction.
    enum Command: Equatable { case skip, like, boost }

    @State private var dragOffset: CGSize = .zero
    @State private var isDismissing = false

    // Reduce Motion (IOS-COMPLY-003): drop the card rotation and the springy
    // overshoot for users who opt out of motion. The drag itself is direct
    // manipulation (not suppressed); only the decorative tilt/spring is.
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    /// Up to 3 cards; deeper cards aren't rendered to keep the layer count low.
    private var visibleItems: ArraySlice<SwipeItem> {
        items.prefix(3)
    }

    /// Vertical room reserved below cards for the behind-card stack peek.
    private static let backCardPeekRoom: CGFloat = 32

    /// Card aspect ratio (width / height).
    private static let cardAspectRatio: CGFloat = 0.7

    /// Shrink the card so rotation overshoot stays inside the deck area.
    private static let rotationSafetyScale: CGFloat = 0.86

    private var cardWidth: CGFloat {
        guard containerWidth > 0 else { return 0 }
        return containerWidth * Self.rotationSafetyScale
    }

    private var cardHeight: CGFloat {
        guard containerHeight > 0 else { return 0 }
        let availableH = max(0, containerHeight - Self.backCardPeekRoom)
        let fromHeight = availableH * Self.cardAspectRatio
        // Use whichever axis is the binding constraint.
        return (min(cardWidth, fromHeight * Self.rotationSafetyScale)) / Self.cardAspectRatio
    }

    var body: some View {
        ZStack {
            ForEach(Array(visibleItems.enumerated()), id: \.element.id) { index, item in
                SwipeCard(
                    item: item,
                    dragOffset: index == 0 ? dragOffset : .zero,
                    isTopCard: index == 0
                )
                .frame(width: cardWidth, height: cardHeight)
                .scaleEffect(scale(for: index))
                .offset(y: stackYOffset(for: index))
                .offset(index == 0 ? dragOffset : .zero)
                .rotationEffect(index == 0 && !reduceMotion ? .degrees(rotationDegrees) : .zero)
                .zIndex(Double(visibleItems.count - index))
                .animation(
                    reduceMotion
                        ? .linear(duration: 0)
                        : .interactiveSpring(response: 0.32, dampingFraction: 0.78),
                    value: dragOffset
                )
                .gesture(dragGesture(for: item), including: index == 0 ? .gesture : .none)
                .onTapGesture { if index == 0 { onTap(item) } }
                .allowsHitTesting(index == 0 && !isDismissing)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        // Drive button taps through the same animated commit path as a swipe.
        .onChange(of: command) { _, newValue in
            guard let newValue else { return }
            switch newValue {
            case .skip: programmaticSkip()
            case .like: programmaticLike()
            case .boost: programmaticBoost()
            }
            command = nil
        }
    }

    // MARK: - Stack geometry

    private func scale(for index: Int) -> CGFloat {
        // Top card unscaled; behind cards 4% smaller per layer.
        max(0.88, 1.0 - CGFloat(index) * 0.04)
    }

    private func stackYOffset(for index: Int) -> CGFloat {
        // Behind cards peek out below the front card.
        CGFloat(index) * 14
    }

    private var rotationDegrees: Double {
        // 1 degree per 18pt of horizontal drag, capped at ±18°.
        let raw = Double(dragOffset.width) / 18
        return max(-18, min(18, raw))
    }

    // MARK: - Gesture

    private func dragGesture(for item: SwipeItem) -> some Gesture {
        DragGesture()
            .onChanged { value in
                guard !isDismissing else { return }
                dragOffset = value.translation
            }
            .onEnded { value in
                guard !isDismissing else { return }
                let h = value.translation.width
                let v = value.translation.height
                let velocity = value.predictedEndTranslation

                // Up-swipe wins over horizontal when it's clearly upward
                // and the vertical magnitude beats the horizontal one.
                let isVertical = abs(v) > abs(h) && v < 0
                let upCommitted = isVertical &&
                    (-v > SwipeCard.commitThreshold || -velocity.height > 320)
                let rightCommitted = !isVertical &&
                    (h > SwipeCard.commitThreshold || velocity.width > 320)
                let leftCommitted = !isVertical &&
                    (-h > SwipeCard.commitThreshold || -velocity.width > 320)

                if upCommitted {
                    commit(item: item, direction: .up, action: onBoost)
                } else if rightCommitted {
                    commit(item: item, direction: .right, action: onLike)
                } else if leftCommitted {
                    commit(item: item, direction: .left, action: onSkip)
                } else {
                    // Spring back (linear snap under Reduce Motion).
                    withAnimation(reduceMotion ? .linear(duration: 0.1) : .spring(response: 0.4, dampingFraction: 0.7)) {
                        dragOffset = .zero
                    }
                }
            }
    }

    private enum CommitDirection { case left, right, up }

    private func commit(item: SwipeItem, direction: CommitDirection, action: @escaping (SwipeItem) -> Void) {
        isDismissing = true
        let target: CGSize
        switch direction {
        case .left: target = CGSize(width: -700, height: dragOffset.height)
        case .right: target = CGSize(width: 700, height: dragOffset.height)
        case .up: target = CGSize(width: dragOffset.width, height: -900)
        }

        switch direction {
        case .left, .right: HapticFeedback.shared.medium()
        case .up: HapticFeedback.shared.premiumUnlock()
        }

        // Reduce Motion: minimize the fly-off travel time (the card must still
        // leave, but we don't draw out the long kinetic sweep).
        withAnimation(.easeOut(duration: reduceMotion ? 0.12 : 0.28)) {
            dragOffset = target
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + (reduceMotion ? 0.12 : 0.28)) {
            action(item)
            dragOffset = .zero
            isDismissing = false
        }
    }

    // MARK: - Programmatic commits

    /// Triggers a left-swipe dismiss as if the user tapped the Skip button.
    func programmaticSkip() { triggerProgrammatic(.left) }
    func programmaticLike() { triggerProgrammatic(.right) }
    func programmaticBoost() { triggerProgrammatic(.up) }

    private func triggerProgrammatic(_ direction: CommitDirection) {
        guard !isDismissing, let item = items.first else { return }
        let action: (SwipeItem) -> Void
        switch direction {
        case .left: action = onSkip
        case .right: action = onLike
        case .up: action = onBoost
        }
        commit(item: item, direction: direction, action: action)
    }
}

File: android/app/src/main/java/com/desmoines/aipulse/ui/screens/discover/SwipeCardStack.kt
package com.desmoines.aipulse.ui.screens.discover

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.FastOutLinearInEasing
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.input.pointer.util.VelocityTracker
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.dp
import com.desmoines.aipulse.data.model.SwipeItem
import com.desmoines.aipulse.util.rememberHapticPerformer
import com.desmoines.aipulse.util.rememberShouldReduceAnimations
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlin.math.abs

/** A button-driven swipe, mapped to the matching commit direction. */
enum class SwipeCommand { SKIP, LIKE, BOOST }

private enum class CommitDirection { LEFT, RIGHT, UP }

/**
 * Tinder-style swipe deck. Renders the top 3 cards as a peek stack and exposes
 * a callback per commit direction; the parent ViewModel is the source of truth
 * for which item is on top. This view owns only the transient drag offset for
 * the in-flight gesture. Mirrors iOS `SwipeCardStack`.
 *
 * Action-bar buttons drive the same animated fly-off via [command].
 */
@Composable
fun SwipeCardStack(
    items: List<SwipeItem>,
    onLike: (SwipeItem) -> Unit,
    onSkip: (SwipeItem) -> Unit,
    onBoost: (SwipeItem) -> Unit,
    onTap: (SwipeItem) -> Unit,
    modifier: Modifier = Modifier,
    command: SwipeCommand? = null,
    onCommandConsumed: () -> Unit = {},
) {
    val density = LocalDensity.current
    val reduceMotion = rememberShouldReduceAnimations()
    val haptic = rememberHapticPerformer()
    val scope = rememberCoroutineScope()

    val offsetX = remember { Animatable(0f) }
    val offsetY = remember { Animatable(0f) }
    var isDismissing by remember { mutableStateOf(false) }

    val commitThresholdPx = with(density) { 110.dp.toPx() }
    val velocityThresholdPx = with(density) { 320.dp.toPx() }
    val flyX = with(density) { 700.dp.toPx() }
    val flyY = with(density) { 900.dp.toPx() }

    suspend fun settle(targetX: Float, targetY: Float, durationMs: Int) = coroutineScope {
        launch { offsetX.animateTo(targetX, tween(durationMs, easing = FastOutLinearInEasing)) }
        launch { offsetY.animateTo(targetY, tween(durationMs, easing = FastOutLinearInEasing)) }
    }

    fun commit(item: SwipeItem, direction: CommitDirection, action: (SwipeItem) -> Unit) {
        if (isDismissing) return
        isDismissing = true
        when (direction) {
            CommitDirection.UP -> haptic.premiumUnlock()
            else -> haptic.medium()
        }
        val targetX = when (direction) {
            CommitDirection.LEFT -> -flyX
            CommitDirection.RIGHT -> flyX
            CommitDirection.UP -> offsetX.value
        }
        val targetY = if (direction == CommitDirection.UP) -flyY else offsetY.value
        val dur = if (reduceMotion) 120 else 280
        scope.launch {
            settle(targetX, targetY, dur)
            action(item)
            offsetX.snapTo(0f)
            offsetY.snapTo(0f)
            isDismissing = false
        }
    }

    fun springBack() {
        scope.launch {
            if (reduceMotion) {
                launch { offsetX.animateTo(0f, tween(100)) }
                offsetY.animateTo(0f, tween(100))
            } else {
                launch { offsetX.animateTo(0f, spring(dampingRatio = 0.7f, stiffness = Spring.StiffnessMedium)) }
                offsetY.animateTo(0f, spring(dampingRatio = 0.7f, stiffness = Spring.StiffnessMedium))
            }
        }
    }

    // Button-driven commits run the same animated path as a gesture swipe.
    LaunchedEffect(command) {
        val cmd = command ?: return@LaunchedEffect
        val top = items.firstOrNull()
        if (top != null && !isDismissing) {
            when (cmd) {
                SwipeCommand.SKIP -> commit(top, CommitDirection.LEFT, onSkip)
                SwipeCommand.LIKE -> commit(top, CommitDirection.RIGHT, onLike)
                SwipeCommand.BOOST -> commit(top, CommitDirection.UP, onBoost)
            }
        }
        onCommandConsumed()
    }

    BoxWithConstraints(modifier = modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        val cardWidth = maxWidth * 0.92f
        val cardHeight = maxHeight - 28.dp
        val visible = items.take(3)

        // Draw back-to-front so the top card (index 0) renders last / on top.
        for (index in visible.indices.reversed()) {
            val item = visible[index]
            val isTop = index == 0
            key(item.id) {
                val scale = (1f - index * 0.04f).coerceAtLeast(0.88f)
                val stackOffsetY = with(density) { (index * 14).dp.toPx() }

                var cardModifier = Modifier
                    .size(cardWidth, cardHeight)
                    .graphicsLayer {
                        scaleX = scale
                        scaleY = scale
                        translationX = if (isTop) offsetX.value else 0f
                        translationY = stackOffsetY + if (isTop) offsetY.value else 0f
                        rotationZ = if (isTop && !reduceMotion) {
                            val dpOffset = offsetX.value / density.density
                            (dpOffset / 18f).coerceIn(-18f, 18f)
                        } else 0f
                    }

                if (isTop) {
                    cardModifier = cardModifier
                        .pointerInput(item.id) {
                            detectTapGestures(onTap = { if (!isDismissing) onTap(item) })
                        }
                        .pointerInput(item.id) {
                            val tracker = VelocityTracker()
                            detectDragGestures(
                                onDragStart = { tracker.resetTracking() },
                                onDrag = { change, drag ->
                                    if (isDismissing) return@detectDragGestures
                                    tracker.addPosition(change.uptimeMillis, change.position)
                                    scope.launch {
                                        offsetX.snapTo(offsetX.value + drag.x)
                                        offsetY.snapTo(offsetY.value + drag.y)
                                    }
                                    change.consume()
                                },
                                onDragCancel = { springBack() },
                                onDragEnd = {
                                    if (isDismissing) return@detectDragGestures
                                    val velocity = tracker.calculateVelocity()
                                    val h = offsetX.value
                                    val v = offsetY.value
                                    val isVertical = abs(v) > abs(h) && v < 0
                                    val up = isVertical &&
                                        (-v > commitThresholdPx || -velocity.y > velocityThresholdPx)
                                    val right = !isVertical &&
                                        (h > commitThresholdPx || velocity.x > velocityThresholdPx)
                                    val left = !isVertical &&
                                        (-h > commitThresholdPx || -velocity.x > velocityThresholdPx)
                                    when {
                                        up -> commit(item, CommitDirection.UP, onBoost)
                                        right -> commit(item, CommitDirection.RIGHT, onLike)
                                        left -> commit(item, CommitDirection.LEFT, onSkip)
                                        else -> springBack()
                                    }
                                },
                            )
                        }
                }

                SwipeCard(
                    item = item,
                    modifier = cardModifier,
                    dragOffsetX = if (isTop) offsetX.value else 0f,
                    dragOffsetY = if (isTop) offsetY.value else 0f,
                    isTopCard = isTop,
                )
            }
        }
    }
}
